core/src/main/scala/kafka/coordinator/transaction/TransactionStateManager.scala

```diff
@@ -108,8 +108,59 @@ class TransactionStateManager(brokerId: Int,
     }
   }
 
-  def enablePidExpiration() {
-    // TODO: add producer id expiration logic
+  def enableTransactionalIdExpiration() {
+    scheduler.schedule("transactionalId-expiration", () => {
+      inLock(stateLock) {
+        val now = time.milliseconds()
+
+        val transactionalIdByPartition: Map[Int, mutable.Iterable[String]] = transactionMetadataCache.flatMap { case (_, entry) =>
+          entry.metadataPerTransactionalId.filter { case (_, txnMetadata) =>
+            txnMetadata.txnLastUpdateTimestamp <= now - config.transactionalIdExpirationMs
+          }.map { case (transactionalId, _) =>
+            transactionalId
+          }
+        }.groupBy { transactionalId =>
+          partitionFor(transactionalId)
+        }
+
+        val recordsPerPartition = transactionalIdByPartition
```

Comment:
@dguy By `Dead` I meant to indicate that the txnId has been dead, not the ongoing transaction since it should have long-been aborted or completed. The problem is that when we are writing the tombstone into the txn log we did not set any pending state in the txn metadata object itself to fence any future request on this txnId while the append is being replicated. So it is possible that after we have written the tombstone but the delayed produce has not been completed, an `addPartition` is received with the txnId.

BTW I think if the current ongoing transaction of the transactionId has complete committed or aborted, its metadata's state should be in CompleteCommit or CompleteAbort, and `Empty` should only happen when this txnId has only been registered from an `initPid` but no txns have ever been started yet, i.e. no `addPartitions` have been received yet.